### ISSUE TITLE
`semi` `omitAfterFunctionBody` option

### ISSUE THREAD

**Issue by Y-Less** [eslint/eslint#14105]
**What rule do you want to change?**

`semi`

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option (`omitAfterFunctionBody` or similar with `always`).

**Please provide some example code that this change will affect:**

```js
function func() {
} // No semi-colon required.

// Refactored slightly:

const func = () => {
} // Suddenly needs a semi-colon.
```

**What does the rule currently do for this code?**

When semi-colons are required (not debating why), there are warnings for missing ones after some, but not all, functions, which is inconsistent.  This means when you slightly change the style at the start of the function for one reason or another (typing, hoisting, binding, etc) you need to change the end of the declaration as well.

**What will the rule do after it's changed?**

Not give a warning/error after function bodies, even when they are in an assignment rather than a function statement.  `beforeStatementContinuationChars` can be used in this case to still give warnings/errors when it matters.  `function` and `const` should be handled as closely as possible in as many cases as possible.

**Are you willing to submit a pull request to implement this change?**

Yes.

**Comment by Y-Less** [comment]
I kept mentioning `=>`/`function`/`const` etc, but really it's for function declaration statements, and function declaration assignments, since this would also apply in cases like class methods and other places

```js
class Class {
  method1() {
  }

  method2 = () => {
  }
}
```

They should be consistent as well.

**Comment by mdjermanovic** [comment]
I'm closing this for the reasons described in the other issue https://github.com/eslint/eslint/issues/14104#issuecomment-778854359

> When semi-colons are required (not debating why), there are warnings for missing ones after some, but not all, functions, which is inconsistent.

`semi` with the option `"always"` requires semi-colons where they're required by the language (or automatically inserted by ASI if omitted).

[LexicalDeclaration](https://tc39.es/ecma262/#prod-LexicalDeclaration) syntax requires `;` at the end.

[FunctionDeclaration](https://tc39.es/ecma262/#prod-FunctionDeclaration) does not have `;` at the end. Semi-colon that comes after a function declaration is treated as an additional (empty) statement that is unrelated to the function declaration.
